libcontainer_dyno_driver.go

```diff
@@ -0,0 +1,554 @@
+	ctx.ex.cmd = cmd
+	return cmd
+}
+
+func (ctx *containerInit) startCallback() {
+	//TODO: log("Starting process web.1 with command `...`")
+	close(ctx.ex.waitStartup)
+	//child process is already running, it's safe to close the parent's read
+	//side of the pipe
+	ctx.configPipe.Close()
+}
+
+type LibContainerInitDriver struct{}
+
+func (dd *LibContainerInitDriver) Build(*Release) error {
+	// noop
+	return nil
+}
+
+// Start acts as PID=1 inside a container spawned by libcontainer, doing the
+// required setup and re-exec'ing as the abspath driver
+// TODO: drop privileges (setuid)
+func (dd *LibContainerInitDriver) Start(ex *Executor) error {
+	configPipe := os.NewFile(4, "configPipe")
+	var container libcontainer.Config
+	decoder := gob.NewDecoder(configPipe)
+	if err := decoder.Decode(&container); err != nil {
+		configPipe.Close()
+		return err
+	}
+	configPipe.Close()
+
+	dynoEnv := make(map[string]string, len(container.Env))
+	for _, entry := range container.Env {
+		pieces := strings.SplitN(entry, "=", 2)
+		dynoEnv[pieces[0]] = pieces[1]
+	}
+	as := AppSerializable{
+		Version: ex.Release.version,
+		Env:     dynoEnv,
+		Slug:    ex.Release.slugURL,
+		Stack:   ex.Release.stack,
+		Processes: []FormationSerializable{
+			{
+				FArgs:     ex.Args,
+				FQuantity: 1,
+				FType:     ex.ProcessType,
+			},
+		},
+	}
+	args := []string{
+		"/tmp/hsup", "-d", "abspath", "-a", ex.Release.appName,
+		"--oneshot", "--start-number=" + ex.ProcessID,
+		"start", ex.ProcessType,
+	}
+	container.Env = []string{
+		"HSUP_SKIP_BUILD=TRUE",
+		"HSUP_CONTROL_GOB=" + as.ToBase64Gob(),
+	}
+	// TODO: clean up /tmp/hsup and /tmp/slug.tgz after abspath reads them
+	return namespaces.Init(
+		&container, container.RootFs, "",
+		os.NewFile(3, "controlPipe"), args,
+	)
+}
+
+func (dd *LibContainerInitDriver) Stop(*Executor) error {
+	panic("this should never be called")
+	return nil
+}
+
+func (dd *LibContainerInitDriver) Wait(*Executor) *ExitStatus {
+	// this should be unreachable, but in case it is called, sleep forever:
+	select {}
+	return nil
+}
+
+func containerConfig(
+	containerUUID, dataPath string, env []string,
+) *libcontainer.Config {
+	return &libcontainer.Config{
+		MountConfig: &libcontainer.MountConfig{
+			MountLabel: containerUUID,
+			PivotDir:   "/tmp",
+			Mounts: []*mount.Mount{
+				{
+					Type:        "bind",
+					Destination: "/app",
+					Writable:    true,
+					Source:      filepath.Join(dataPath, "app"),
+				},
+				{
+					Type:        "bind",
+					Destination: "/tmp",
+					Writable:    true,
+					Source:      filepath.Join(dataPath, "tmp"),
+				},
+				{
+					Type:        "bind",
+					Destination: "/var/tmp",
+					Writable:    true,
+					Source: filepath.Join(
+						dataPath,
+						"var", "tmp",
+					),
+				},
+				{
+					Type:        "bind",
+					Writable:    false,
+					Destination: "/etc/resolv.conf",
+					Source:      "/etc/resolv.conf",
+				},
+			},
+			DeviceNodes: []*devices.Device{
```

Comment:
Could replace this with `devices.DefaultSimpleDevices`, though it could change unexpectedly.